File: demo/src/main/java/com/devbrackets/android/playlistcoredemo/helper/BaseMediaApi.java
package com.devbrackets.android.playlistcoredemo.helper;

import android.media.MediaPlayer;

import com.devbrackets.android.playlistcore.api.MediaPlayerApi;
import com.devbrackets.android.playlistcore.listener.MediaStatusListener;
import com.devbrackets.android.playlistcoredemo.data.MediaItem;

import org.jetbrains.annotations.NotNull;

public abstract class BaseMediaApi implements MediaPlayerApi<MediaItem>,
        MediaPlayer.OnPreparedListener,
        MediaPlayer.OnCompletionListener,
        MediaPlayer.OnErrorListener,
        MediaPlayer.OnSeekCompleteListener,
        MediaPlayer.OnBufferingUpdateListener {

    protected boolean prepared;
    protected int bufferPercent;

    protected MediaStatusListener<MediaItem> mediaStatusListener;

    @Override
    public void setMediaStatusListener(@NotNull MediaStatusListener<MediaItem> listener) {
        mediaStatusListener = listener;
    }

    @Override
    public void onCompletion(MediaPlayer mp) {
        if (mediaStatusListener != null) {
            mediaStatusListener.onCompletion(this);
        }
    }

    @Override
    public void onPrepared(MediaPlayer mp) {
        prepared = true;

        if (mediaStatusListener != null) {
            mediaStatusListener.onPrepared(this);
        }
    }

    @Override
    public boolean onError(MediaPlayer mp, int what, int extra) {
        return mediaStatusListener != null && mediaStatusListener.onError(this);
    }

    @Override
    public void onSeekComplete(MediaPlayer mp) {
        if (mediaStatusListener != null) {
            mediaStatusListener.onSeekComplete(this);
        }
    }

    @Override
    public void onBufferingUpdate(MediaPlayer mp, int percent) {
        bufferPercent = percent;

        if (mediaStatusListener != null) {
            mediaStatusListener.onBufferingUpdate(this, percent);
        }
    }
}

File: demo/src/main/java/com/devbrackets/android/playlistcoredemo/helper/BaseMediaApi.kt
package com.devbrackets.android.playlistcoredemo.helper

import com.devbrackets.android.exomedia.listener.*
import com.devbrackets.android.playlistcore.api.MediaPlayerApi
import com.devbrackets.android.playlistcore.listener.MediaStatusListener
import com.devbrackets.android.playlistcoredemo.data.MediaItem

abstract class BaseMediaApi :
  MediaPlayerApi<MediaItem>,
  OnPreparedListener,
  OnCompletionListener,
  OnErrorListener,
  OnSeekCompletionListener,
  OnBufferUpdateListener
{
  protected var prepared = false
  protected var bufferPercent = 0
  protected var statusListener: MediaStatusListener<MediaItem>? = null

  override fun setMediaStatusListener(listener: MediaStatusListener<MediaItem>) {
    statusListener = listener
  }

  override fun onCompletion() {
    statusListener?.onCompletion(this)
  }

  override fun onError(e: Exception): Boolean {
    return statusListener?.onError(this) == true
  }

  override fun onPrepared() {
    prepared = true
    statusListener?.onPrepared(this)
  }

  override fun onSeekComplete() {
    statusListener?.onSeekComplete(this)
  }

  override fun onBufferingUpdate(percent: Int) {
    bufferPercent = percent
    statusListener?.onBufferingUpdate(this, percent)
  }
}